zrml/neo-swaps/src/math.rs

```diff
@@ -278,5 +324,47 @@ mod detail {
             let result: Fixed = exp(Fixed::checked_from_num(value).unwrap(), neg).unwrap();
             assert_eq!(result, expected);
         }
+
+        #[test]
+        fn overflow() {
+            let reserve = 10_000 * BASE;
+            let liquidity = 3_333_333_333_333;
+            let amount_in = 5 * liquidity;
+            let amount_out =
+                calculate_swap_amount_out_for_buy(reserve, amount_in, liquidity).unwrap();
+            assert_eq!(amount_out, 99_977_464_168_502);
+            let sum_of_prices = calculate_spot_price(18_977_157_268_533, liquidity).unwrap()
+                + calculate_spot_price(18_977_157_268_533, liquidity).unwrap()
+                + calculate_spot_price(22_535_831_498, liquidity).unwrap();
+            assert_eq!(sum_of_prices, BASE);
+        }
+
+        #[test]
+        fn overflow2() {
+            let reserve = 10_000 * BASE;
+            let liquidity = 4_342_944_819_032;
+            let amount_in = 10 * liquidity;
+            let amount_out =
+                calculate_swap_amount_out_for_buy(reserve, amount_in, liquidity).unwrap();
+            assert_eq!(amount_out, 99_999_802_826_134);
+            let sum_of_prices = calculate_spot_price(46_439_748_147_833, liquidity).unwrap()
+                + calculate_spot_price(46_439_748_146_964, liquidity).unwrap()
+                + calculate_spot_price(197_173_865, liquidity).unwrap();
+            assert_approx!(sum_of_prices, BASE, 1);
+        }
+
+        #[test]
+        fn overflow3() {
+            let reserve = 10_000 * BASE;
+            let liquidity = 4_342_944_819_032;
+            let amount_in = BASE / 100;
+            let amount_out =
+                calculate_swap_amount_out_for_buy(reserve, amount_in, liquidity).unwrap();
+            assert_eq!(amount_out, 53_622_125_748_008); // Actual result: 53622125747983
+            let sum_of_prices = calculate_spot_price(3_010_399_957_508, liquidity).unwrap()
+                + calculate_spot_price(3_010_399_956_639, liquidity).unwrap()
+                + calculate_spot_price(46_377_874_251_992, liquidity).unwrap();
+            assert_approx!(sum_of_prices, BASE, 1);
+        }
     }
 }
```

Comment:
It seems to me test coverage is relatively low in this file.

Reply:
Fixed over this series of commits: 

52742d210bf61f325707a047fa662d3e67c70734
a924613ac62c33fc93f24cc8e77f88da31c1b5c9
286f49df753972b165793b2aefedc60ec1fc5843
1664dcbafc64ed9815bdaaa61031d471b15a5234
06e3888feba99dfae4a69d1a7464f31fae10819a